### ISSUE TITLE
Cosine metric is not a valid metric not even a pseudo metric (even when restricted to unit vectors/n-sphere)

### ISSUE THREAD

**Issue by franciscovargas** [scikit-tda/ripser.py#122]
Hi ,

The ripser.py.ripser class has a flag `metric` (https://github.com/scikit-tda/ripser.py/blob/ff3f017e451cbf1d9c31acaa55ab8510826d960c/ripser/ripser.py#L149) that enabels a cosine metric. For a third party user you expect this to be a valid metric thus one would guess it is the angular distance https://en.wikipedia.org/wiki/Cosine_similarity#Angular_distance_and_similarity which can be shown to be a metric .

 However if one digs in the code we can see that ripser.py uses an sklearn function to calculate the distance metrics https://github.com/scikit-tda/ripser.py/blob/ff3f017e451cbf1d9c31acaa55ab8510826d960c/ripser/ripser.py#L34 . Said sklearn function implements the cosine distance `1 - <a,b>/||a|| ||b||` which is not even a pseudometric as it violates the triangle inequality.  This can be seen here https://github.com/scikit-learn/scikit-learn/blob/15a949460dbf19e5e196b8ef48f9712b72a3b3c3/sklearn/metrics/pairwise.py#L807 . This seems like a bug, violating the triangle inequality feels like too strong of a break in the conditions of a metric .

**Comment by ctralie** [comment]
You're right, but this seems like more of a problem with scikit-learn, no?
How would you propose we fix this, given that there are plenty of other
useful metrics there, and we wrapped around scikit-learn to avoid
re-inventing the wheel?  Shouldn't you take this issue up with scikit-learn
instead?
